source/fswatch.d

```diff
@@ -321,24 +345,24 @@ struct FileWatch
 						if (info.cookie == cookie)
 						{
 							events ~= FileChangeEvent(FileChangeEventType.rename,
-									fromFilename, fileName);
+									fromFilename, relativeFilename);
 						}
 						else
-							events ~= FileChangeEvent(FileChangeEventType.create, fileName);
+							events ~= FileChangeEvent(FileChangeEventType.create, relativeFilename);
 						cookie = 0;
 					}
 					if ((info.mask & IN_DELETE_SELF) != 0 || (info.mask & IN_MOVE_SELF) != 0)
 					{
 						if (fd)
 						{
-							inotify_rm_watch(fd, wd);
-							close(fd);
-							fd = wd = 0;
+							inotify_rm_watch(fd, info.wd);
+							wds[info.wd] = false;
 						}
-						events ~= FileChangeEvent(FileChangeEventType.removeSelf, ".");
+						if (directoryMap[info.wd] == path)
+							events ~= FileChangeEvent(FileChangeEventType.removeSelf, path);

```

Comment:
this shouldn't be removeSelf for anything else than the path itself

Reply:
This is the reason for the check. As you can see from the test routine, it sends a "remove" event on subdirectoris